From afd823b21c544eb4764d9e38f2f15e5a4f50ab7a Mon Sep 17 00:00:00 2001
From: David Henningsson <david.henningsson@canonical.com>
Date: Wed, 17 Apr 2013 16:07:32 +0200
Subject: [PATCH] inotify-wrapper: Quit daemon if pid file is removed

XDG_RUNTIME_DIR is deleted on logout (according to xdg spec), and
that's where we keep our pid file and native protocol socket.

When these two files are deleted, it makes no sense to run anymore,
so quit. Hanging around is not only sloppy, but can also be harmful:
e g, if esd compat module is loaded, it will bind to the esd address.
When a new pulseaudio process is created on next login of the same
user, pulseaudio cannot start because the esd address is occupied.

Buglink: https://bugs.launchpad.net/bugs/1167192
Signed-off-by: David Henningsson <david.henningsson@canonical.com>
---
 src/Makefile.am                 |    1 +
 src/daemon/main.c               |   20 +++++++-
 src/pulsecore/inotify-wrapper.c |  103 +++++++++++++++++++++++++++++++++++++++
 src/pulsecore/inotify-wrapper.h |   13 +++++
 src/pulsecore/pid.c             |    5 ++
 src/pulsecore/pid.h             |    1 +
 6 files changed, 141 insertions(+), 2 deletions(-)
 create mode 100644 src/pulsecore/inotify-wrapper.c
 create mode 100644 src/pulsecore/inotify-wrapper.h

Index: pulseaudio/src/Makefile.am
===================================================================
--- pulseaudio.orig/src/Makefile.am
+++ pulseaudio/src/Makefile.am
@@ -938,6 +938,7 @@ libpulsecore_@PA_MAJORMINOR@_la_SOURCES
 		pulsecore/core-subscribe.c pulsecore/core-subscribe.h \
 		pulsecore/core.c pulsecore/core.h \
 		pulsecore/hook-list.c pulsecore/hook-list.h \
+		pulsecore/inotify-wrapper.c pulsecore/inotify-wrapper.h \
 		pulsecore/ltdl-helper.c pulsecore/ltdl-helper.h \
 		pulsecore/modargs.c pulsecore/modargs.h \
 		pulsecore/modinfo.c pulsecore/modinfo.h \
Index: pulseaudio/src/daemon/main.c
===================================================================
--- pulseaudio.orig/src/daemon/main.c
+++ pulseaudio/src/daemon/main.c
@@ -73,6 +73,7 @@
 #include <pulsecore/core-rtclock.h>
 #include <pulsecore/core-scache.h>
 #include <pulsecore/core.h>
+#include <pulsecore/inotify-wrapper.h>
 #include <pulsecore/module.h>
 #include <pulsecore/cli-command.h>
 #include <pulsecore/log.h>
@@ -332,6 +333,15 @@ static char *check_configured_address(vo
     return default_server;
 }
 
+static bool valid_pid_file = false;
+static void pid_file_deleted(void *userdata)
+{
+    pa_core *c = userdata;
+    pa_log_info("Our pid file has been deleted (probably due to session logout), quitting...");
+    valid_pid_file = false;
+    pa_core_exit(c, true, 0);
+}
+
 #ifdef HAVE_DBUS
 static pa_dbus_connection *register_dbus_name(pa_core *c, DBusBusType bus, const char* name) {
     DBusError error;
@@ -374,7 +384,6 @@ int main(int argc, char *argv[]) {
     char *s;
     char *configured_address;
     int r = 0, retval = 1, d = 0;
-    bool valid_pid_file = false;
     bool ltdl_init = false;
     int n_fds = 0, *passed_fds = NULL;
     const char *e;
@@ -382,6 +391,7 @@ int main(int argc, char *argv[]) {
     int daemon_pipe[2] = { -1, -1 };
     int daemon_pipe2[2] = { -1, -1 };
 #endif
+    pa_inotify *pid_monitor = NULL;
     int autospawn_fd = -1;
     bool autospawn_locked = false;
 #ifdef HAVE_DBUS
@@ -1022,6 +1032,9 @@ int main(int argc, char *argv[]) {
         goto finish;
     }
 
+    if (valid_pid_file)
+        pid_monitor = pa_inotify_start(pa_pid_file_name(), c, pid_file_deleted, c);
+
     c->default_sample_spec = conf->default_sample_spec;
     c->alternate_sample_rate = conf->alternate_sample_rate;
     c->default_channel_map = conf->default_channel_map;
@@ -1162,6 +1175,9 @@ finish:
         pa_autospawn_lock_done(false);
     }
 
+    if (pid_monitor)
+        pa_inotify_stop(pid_monitor);
+
     if (c) {
         /* Ensure all the modules/samples are unloaded when the core is still ref'ed,
          * as unlink callback hooks in modules may need the core to be ref'ed */
Index: pulseaudio/src/pulsecore/inotify-wrapper.c
===================================================================
--- /dev/null
+++ pulseaudio/src/pulsecore/inotify-wrapper.c
@@ -0,0 +1,103 @@
+#ifdef HAVE_CONFIG_H
+#include <config.h>
+#endif
+
+#include <fcntl.h>
+#include <sys/inotify.h>
+#include <errno.h>
+
+#include "inotify-wrapper.h"
+#include <pulse/mainloop.h>
+#include <pulsecore/core-util.h>
+#include <pulsecore/core-error.h>
+
+struct pa_inotify {
+    char *filename;
+    void *callback_data;
+    pa_inotify_cb callback;
+    int fd;
+    pa_io_event *io_event;
+    pa_core *core;
+};
+
+
+static void inotify_cb(
+        pa_mainloop_api *a,
+        pa_io_event *e,
+        int fd,
+        pa_io_event_flags_t events,
+        void *userdata) {
+
+    struct {
+        struct inotify_event event;
+        char name[NAME_MAX+1];
+    } buf;
+
+    pa_inotify *i = userdata;
+    int pid_fd;
+
+    pa_assert(i);
+
+    if (pa_read(fd, &buf, sizeof(buf), NULL) < (int) sizeof(buf.event))
+        pa_log_warn("inotify did not read a full event.");
+    else
+        pa_log_debug("inotify callback, event mask: 0x%x", (int) buf.event.mask);
+
+    pid_fd = pa_open_cloexec(i->filename, O_RDONLY
+#ifdef O_NOFOLLOW
+                       |O_NOFOLLOW
+#endif
+                       , S_IRUSR);
+
+    if (pid_fd < 0) {
+        if (i->callback)
+            i->callback(i->callback_data);
+    } else
+        pa_close(pid_fd);
+}
+
+
+pa_inotify *pa_inotify_start(const char *filename, void *userdata, pa_inotify_cb cb, pa_core *core) {
+
+    pa_inotify *i = pa_xnew0(pa_inotify, 1);
+    pa_assert(i);
+
+    i->core = core;
+    pa_core_ref(core);
+
+    i->filename = pa_xstrdup(filename);
+    i->callback_data = userdata;
+    i->callback = cb;
+    i->fd = inotify_init1(IN_CLOEXEC|IN_NONBLOCK);
+
+    if (i->fd < 0) {
+        pa_log("inotify_init1() failed: %s", pa_cstrerror(errno));
+        pa_inotify_stop(i);
+        return NULL;
+    }
+
+    if (inotify_add_watch(i->fd, filename, IN_DELETE_SELF|IN_MOVE_SELF) < 0) {
+        pa_log("inotify_add_watch() failed: %s", pa_cstrerror(errno));
+        pa_inotify_stop(i);
+        return NULL;
+    }
+
+    pa_assert_se(i->io_event = core->mainloop->io_new(core->mainloop, i->fd, PA_IO_EVENT_INPUT, inotify_cb, i));
+
+    return i;
+}
+
+
+void pa_inotify_stop(pa_inotify *i) {
+
+    pa_assert(i);
+
+    if (i->io_event)
+        i->core->mainloop->io_free(i->io_event);
+    if (i->fd)
+        pa_close(i->fd);
+    pa_xfree(i->filename);
+    if (i->core)
+        pa_core_unref(i->core);
+    pa_xfree(i);
+}
Index: pulseaudio/src/pulsecore/inotify-wrapper.h
===================================================================
--- /dev/null
+++ pulseaudio/src/pulsecore/inotify-wrapper.h
@@ -0,0 +1,13 @@
+#ifndef fooinotifywrapperhfoo
+#define fooinotifywrapperhfoo
+
+#include <pulsecore/core.h>
+
+typedef struct pa_inotify pa_inotify;
+
+typedef void (*pa_inotify_cb)(void *userdata);
+
+pa_inotify *pa_inotify_start(const char *filename, void *userdata, pa_inotify_cb cb, pa_core *c);
+void pa_inotify_stop(pa_inotify *i);
+
+#endif
Index: pulseaudio/src/pulsecore/pid.c
===================================================================
--- pulseaudio.orig/src/pulsecore/pid.c
+++ pulseaudio/src/pulsecore/pid.c
@@ -181,6 +181,11 @@ static int proc_name_ours(pid_t pid, con
 
 }
 
+char *pa_pid_file_name()
+{
+    return pa_runtime_path("pid");
+}
+
 /* Create a new PID file for the current process. */
 int pa_pid_file_create(const char *procname) {
     int fd = -1;
Index: pulseaudio/src/pulsecore/pid.h
===================================================================
--- pulseaudio.orig/src/pulsecore/pid.h
+++ pulseaudio/src/pulsecore/pid.h
@@ -20,6 +20,7 @@
   License along with PulseAudio; if not, see <http://www.gnu.org/licenses/>.
 ***/
 
+char *pa_pid_file_name(void);
 int pa_pid_file_create(const char *procname);
 int pa_pid_file_remove(void);
 int pa_pid_file_check_running(pid_t *pid, const char *procname);
